x-pack/plugins/ml/server/lib/alerts/register_anomaly_detection_alert_type.ts

```diff
@@ -42,10 +62,88 @@ export type AnomalyDetectionAlertContext = AnomalyDetectionAlertBaseContext & {
   topInfluencers?: InfluencerAnomalyAlertDoc[];
 };
 
+export type ExecutorOptions<P extends RuleTypeParams> = RuleExecutorOptions<
+  P,
+  RuleTypeState,
+  {},
+  AnomalyDetectionAlertContext,
+  typeof ANOMALY_SCORE_MATCH_GROUP_ID,
+  MlAnomalyDetectionAlert
+>;
+
 export const ANOMALY_SCORE_MATCH_GROUP_ID = 'anomaly_score_match';
 
 export type AnomalyScoreMatchGroupId = typeof ANOMALY_SCORE_MATCH_GROUP_ID;
 
+export const ANOMALY_DETECTION_AAD_INDEX_NAME = 'ml.anomaly-detection';
+
+const ML_ALERT_NAMESPACE = ALERT_NAMESPACE;
+
+export const ALERT_ANOMALY_DETECTION_JOB_ID = `${ML_ALERT_NAMESPACE}.job_id` as const;
+
+export const ALERT_ANOMALY_SCORE = `${ML_ALERT_NAMESPACE}.anomaly_score` as const;
+export const ALERT_ANOMALY_IS_INTERIM = `${ML_ALERT_NAMESPACE}.is_interim` as const;
+export const ALERT_ANOMALY_TIMESTAMP = `${ML_ALERT_NAMESPACE}.anomaly_timestamp` as const;
+
+export const ALERT_TOP_RECORDS = `${ML_ALERT_NAMESPACE}.top_records` as const;
+export const ALERT_TOP_INFLUENCERS = `${ML_ALERT_NAMESPACE}.top_influencers` as const;
+
+export const ANOMALY_DETECTION_AAD_CONFIG: IRuleTypeAlerts<MlAnomalyDetectionAlert> = {
+  context: ANOMALY_DETECTION_AAD_INDEX_NAME,
+  mappings: {
+    fieldMap: {
+      [ALERT_ANOMALY_DETECTION_JOB_ID]: {
+        type: ES_FIELD_TYPES.KEYWORD,
+        array: false,
+        required: true,
+      },
+      [ALERT_ANOMALY_SCORE]: { type: ES_FIELD_TYPES.DOUBLE, array: false, required: false },
+      [ALERT_ANOMALY_IS_INTERIM]: { type: ES_FIELD_TYPES.BOOLEAN, array: false, required: false },
+      [ALERT_ANOMALY_TIMESTAMP]: { type: ES_FIELD_TYPES.DATE, array: false, required: false },
+      [ALERT_TOP_RECORDS]: {
+        type: ES_FIELD_TYPES.NESTED,
+        array: true,
+        required: false,
+        dynamic: false,
```

Comment:
One thing to understand here is that when we write anomalies to the anomaly results indices we added the partition/by/over fields twice:

1. We add them as `partition_field_name`/`partition_field_value` etc.
2. We add them using their original field names, for example `airline`/`AAL`

In the anomaly results indices we try to map the original field names as well as the predictable ones. This means that you can search anomaly results with a terms search on the `airline` field and get hits.

Setting `dynamic: false` here means that these sub-objects _won't_ be searchable by original field names. If the values have been populated by copying the `_source` of the original anomalies in full then the original fields will be visible in the `_source` of the alerts, but the alerts won't be searchable by these original field names.

Unfortunately, trying to create the original field names in the mappings in the anomaly results indices has turned out to be a very difficult problem, and has resulted in a lot of complicated code and support cases when things go wrong.

Therefore I think it would be best _not_ to try and do the same for alerts. (In other words leave this part of the mappings as it is now.) But it's important to realise that, as a result, it won't be possible to search for alerts for `airline`/`AAL`, but instead it will be necessary to search for `partition_field_name`/`airline` and `partition_field_value`/`AAL`.  This then ties back to the `object` vs `nested` decision, because you can only search for two fields together in arrays of objects if you use the `nested` type.

Reply:
Good point regarding searching by the original field names and allowing dynamic mappings. It only affects `.alerts-ml.anomaly-detection.alerts-default` indices so it shouldn't be an issue. 
To clarify, will we be able to query for  `airline:AAL` with the `object` type?

Reply:
Would it be possible to use the `flattened` mapping to be able to search for `airline:AAL` inside the copy of the original `_source`?

Reply:
I think it's important to know the actual field name. With the `flattened` mapping we can end up with similar field values coming from completely different fields if I understand it correctly. WDYT @droberts195?

Reply:
> To clarify, will we be able to query for `airline:AAL` with the `object` type?

No, only if there was an explicit mapping for `airline` within the `object`. And getting those explicit mappings in is extremely hard for fields that vary by job.

> Would it be possible to use the `flattened` mapping

`flattened` might be ideal if `alert_top_records` was a single object rather than an array. I am not sure how `flattened` will behave with an array of objects though - that would need some experimentation.

One more thing about `flattened` is that it treats all sub-values as keywords. So then it definitely won't be possible to do a search like, "Give me all alerts where one of the top records has a record score greater than 90."

It all comes back to how we expect users to be searching these alerts. If we have a feel for how people will want to search, filter and sort then it will be easier to create mappings that support those use cases (or at least say with certainty that it's impossible to support all of them and document the use cases we _do_ support).

Reply:
Updated type to `object` in https://github.com/elastic/kibana/pull/166349/commits/33d037c43aa592058a9704875dc9ca4ac0b77b46